accepted/2025/rid-specific-tool-packages.md

```diff
@@ -0,0 +1,91 @@
+# RID-Specific .NET Tool Packages
+
+**Owner** [Daniel Plaisted](https://github.com/dsplaisted)
+
+## Summary
+
+This proposal adds the ability for a .NET Tool to have separate packages for each supported OS and architecture.
+
+In .NET, the combination of OS and architecture is represented by a Runtime Identifier (RID).  Today, .NET Tools support native assets, but the native assets for all of the supported Runtime Identifiers need to be included in the same package.  For tools with large native dependencies, or if the entire tool is native (for example via NativeAOT), this multiplies the package download size by the number of supported RIDs.
+
+We would like to add support for RID-specific .NET Tool packages.  This would enable a tool to have a separate package for each supported RID, and when the tool is installed only the package for the correct RID would need to be downloaded.
+
+### Goals
+
+- Tools can create RID-specific packages
+- Installing a tool with RID-specific packages is transparent but only downloads the package for the current RID
+
+### Non-goals
+
+- Allowing a tool to have assets from a mixture of packages, for example have architecture neutral assets in the primary package and native dependencies in RID-specific packages
+
+## Tool manifests
+
+.NET Tools have a tool manifest file named `DotnetToolSettings.xml`.  It is stored alongside the executable tool assets in the `tools\<TargetFramework>\<RuntimeIdentifier>` folder.  Currently for most or all tools, the `RuntimeIdentifier` is `any`.  Here is an example of a current manifest:
+
+```xml
+<DotNetCliTool Version="1">
+  <Commands>
+    <Command Name="dotnet-say" EntryPoint="dotnet-say.dll" Runner="dotnet" />
+  </Commands>
+</DotNetCliTool>
+```
+
+## Design
+
+A tool with RID-specific packages will consist of a single primary package and RID-specific packages for each supported RID.  The primary package will include a tool manifest that lists the RIDs supported by the tool, and the package name and package version of the RID-specific tool package for each one.  The primary package will not include any tool implementation assets or shims.  The RID-specific packages will have the same layout and format as tool packages currently have.  The only difference will be that the NuGet package type will be set to a new `DotnetToolRidPackage` type, in order to prevent tool search results from being cluttered with RID-specific tool packages (the primary package is the one that should show up in the results).
```

Comment:
Will the RID-specific package be validate to have `DotnetToolRidPackage`? Or will any existing package be accepted and treated as a tool package? 

The relationship is only 2 layers deep, right? `1 DotnetTool` to `N DotnetToolRidPackage`, and no `DotnetTool -> DotnetTool -> DotnetToolRidPackage`?

Reply:
The RID-specific package won't have `DotnetToolRidPackage` in its manifest, that's only for the primary package to redirect to the RID-specific packages.  Theoretically a tool package could redirect to any other tool package.  There's currently only layers though, it wouldn't keep following redirects.

We might also decide we want to support tool redirects.  I think that might be useful as part of the [one-shot / dnx / dotnet tool exec work](https://github.com/dotnet/sdk/issues/31103) in order to hide the difference between a tool package name and its command name.  But that's not part of a proposal yet.